In [1]:
import numpy as np

## Top Flange Stress: UG-34

Calculation #1

Top flange material: SS304

d: diameter of the flange, 18 in \
C: 0.25 from sketch (p) of Fig. UG-34 \
P: \
S:
SLArpaas flange thickness: 0.5 in

In [10]:
d = 18
C = 0.25
P = 12.5
S = 20000
E = 1
t = d*np.sqrt(C*P/(S*E))
print( f't = {t}')

t = 0.225


## Ring Flange Stress

We have Flat Face Flanges with Metal-to-Metal Contacnt Outside the Bolt Circle, and therefore should use non-Mandatory Appendix Y, Class 3 Flange Assembly, Figure Y-5.1.3.  (For a large central hole in a flat plate where the hole is greater than 1/2 the flange diameter, the rules of Mandatory Appendix 14 should be used.)

Flange thinkness: 0.5", material: stainless steel 304

### Calculation #2A

The conservative calculation for a ring flange with 18-in OD and one X-diameter hole in the center of the flange.  Wes said it doesn't apply; to check with him.

### Bolt load

### Equivalent hole area

3-in holes x 3 \
1.7-in holes x 2 \
1.2-in holes x 2

In [7]:
# area of a 3" hole
A3 = (1.5**2)*np.pi
# area of a 1.7" hole
A1p7 = (0.85**2)*np.pi
# area of a 1.2" hole
A1p2 = (0.6**2)*np.pi
A = 3*A3 + 2*A1p7 + 2*A1p2
print( f'Area of the 3-in hole: {A3} sq in; area of the 1.7-in hole: {A1p7} sq in; area of the 1.2-in hole: {A1p2} sq in; total area: {A} sq in')

Area of the 3-in hole: 7.0685834705770345 sq in; area of the 1.7-in hole: 2.269800692218625 sq in; area of the 1.2-in hole: 1.1309733552923256 sq in; total area: 28.007298506753006 sq in


In [8]:
# Equivalent diameter
d_eq = np.sqrt(A/np.pi)
print( f'Equivalent to a single hole with the diamter {d_eq} in')

Equivalent to a single hole with the diamter 2.9857997253667232 in


## Openings

### Welded connections:

UG-36 (c)(3)(-a): 
> 3 1/2 in. (89 mm) diameter — in vessel shells or heads with a required minimum thickness of 3/8 in. (10 mm) or less;

The `required minimum thickness` here means the `t` calculated for the top flange stress analysis (calculation #1, based on UG-34).  So (-a) and (-b) together make sense.  Our `t` = 0.225 and therefore we can use (-a).  Our largest openings have the diameter of 3 in, smaller than 3.5 in. 

&rarr; No reinforcement is required for welded connections.

### Distance between openings

UG-36 (c)(3)(-c): 
> no two isolated unreinforced openings, in accordance with (-a) or (-b) above, shall have their centers closer to each other than the sum of their diameters;

Between the three 3-in openings: 8.113 in, 8.113 in, 9 in; all larger than 6 in &rarr; OK

Between the two 1.7-in openings: 7.601 in &rarr; OK

Between the two 1.2-in openings: 2.253 in &rarr; Not OK, requires reinforcement (1); 2.253 in > 1.2 in$\times$1.25 = 1.5 in &rarr; UG-39 (b)(2)

Between the 3-in and 1.2-in openings: 3.536 in, 4.977 in &rarr; Not OK, requires reinforcement (2); 3.536 in > (3 in+1.2 in)/2$\times$1.25 = 2.625 in &rarr; UG-39 (b)(2)

Between the 3-in and 1.7-in openings (the same distance): 4.562 in &rarr; Not OK, requires reinforcement (3); 4.562 in > (3 in+1.7 in)/2$\times$1.25 = 2.9375 in &rarr; UG-39 (b)(2)

Between the 3-in and another 1.7-in openings: 4.707 in, 5.78 in &rarr; OK

### Reinforcement required for openings in flat heads and covers

UG-39 (b)(1)

$f_{r1}$ = 1: $f_{r1}$ is defined in UG-37, and our case corresponds to Figure UG-40, sketch (o)

&rarr; $A = 0.5dt$

$t$ = minimum required thickness of flat head or cover (UG-34)

$d$ = finished diameter of circular opening or finished dimension (chord length at mid- surface of thickness excluding excess thick- ness available for reinforcement) of nonradial opening in the plane under consideration, in. (mm) (UG-37)

UG-39 (b)(2)

> Multiple openings none of which have diameters exceeding one‐half the head diameter and no pair having an average diameter greater than one‐quarter the head diameter may be reinforced individually as required by (1) above when the spacing between any pair of adjacent openings is equal to or greater than twice the average diameter of the pair.
>
> When spacing between adjacent openings is less than twice but equal to or more than 11/4 the average diameter of the pair, the required reinforcement for each opening in the pair, as determined by (1) above, shall be summed together and then distributed such that 50% of the sum is located between the two openings. Spacings of less than 11/4 the average diameter of adjacent openings shall be treated by rules of U-2(g).

In [25]:
# Reinforcement calculations
def UG39_b_2(s, d1, d2, t, t_rel):
    d_avg = (d1+d2)/2
    A = 0.5* d_avg *t
    R = (s - d_avg)*(t_rel -t)
    isReinforced = False
    if R > A:
        isReinforced = True

    return A, R, isReinforced

#### Reinforcement (1)

In [26]:
A, R, isReinforced = UG39_b_2(2.253, 1.2, 1.2, t, 0.5)
print(f'A = {A} sq in, R = {R} sq in, {isReinforced}')

A = 0.135 sq in, R = 0.2895750000000001 sq in, True


#### Reinforcement (2)

In [27]:
A, R, isReinforced = UG39_b_2(3.536, 3, 1.2, t, 0.5)
print(f'A = {A} sq in, R = {R} sq in, {isReinforced}')

A = 0.23625000000000002 sq in, R = 0.39490000000000003 sq in, True


#### Reinforcement (3)

In [28]:
A, R, isReinforced = UG39_b_2(4.562, 3, 1.7, t, 0.5)
print(f'A = {A} sq in, R = {R} sq in, {isReinforced}')

A = 0.264375 sq in, R = 0.6083000000000001 sq in, True
